keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md

```diff
@@ -261,6 +265,38 @@ Unit, integration and E2E tests cover the existing StatefulSet mechanics.
 Additionally, unit and integration tests will be added to cover the
 API validation, behavioral change of StatefulSet with feature gate enabled and disabled.
 
+#### Prerequisite testing updates
+
+[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
+
+##### Unit tests
+
+```
+`k8s.io/kubernetes/pkg/apis/apps/validation`                       `06/07/2022`:    `90.6% of statements` `The tests added for the current feature in this package touches the statefulSet Spec and Status fields. No new tests are needed for promotion to GA`
+`k8s.io/kubernetes/pkg/apis/apps/validation/validation.go:93`:	   `06/07/2022`:		`93.1% of statements`
+`k8s.io/kubernetes/pkg/apis/apps/validation/validation.go:162`:	   `06/07/2022`:		`100% of statements`
+`k8s.io/kubernetes/pkg/apis/apps/validation/validation.go:169`:	   `06/07/2022`:		`94.1% of statements`
+`k8s.io/kubernetes/pkg/apis/apps/validation/validation.go:199`:	   `06/07/2022`:		`100% of statements`
+`k8s.io/kubernetes/pkg/controller/statefulset`:                    `06/07/2022`:    `85.5% of statements` `The tests added for the current feature in this package touches the statefulSet upgrade strategies. No new tests are needed for promotion to GA`
+`k8s.io/kubernetes/pkg/registry/apps/statefulset`:                 `06/07/2022`:    `76.7% of statements` `The tests added for the current feature in this package makes sure that the kubernetes version upgrades won't have any impact on the new fields to the statefulset api when persisting to etcd. No new tests are needed for promotion to GA`
+`k8s.io/kubernetes/pkg/registry/apps/statefulset/strategy.go:95`:	 `06/07/2022`:	  `100.0% of statements`
+`k8s.io/kubernetes/pkg/registry/apps/statefulset/strategy.go:139`: `06/07/2022`:	  `100.0% of statements`
+`k8s.io/kubernetes/pkg/registry/apps/statefulset/strategy.go:223`: `06/07/2022`:	  `100.0% of statements`
+`k8s.io/kubernetes/pkg/registry/apps/statefulset/strategy.go:118`: `06/07/2022`:	  `100.0% of statements`
+```
+
+##### Integration tests
+
+Added integration tests to test availabile replicas when minReadySeconds is set on the statefulset spec.
+
+k8s.io/kubernetes/test/integration/statefulset.TestStatefulSetAvailable: [test grid](https://testgrid.k8s.io/sig-release-master-blocking#integration-master)
+
+##### e2e tests
+
+Following e2e tests are added to statefulsets.
+- StatefulSet MinReadySeconds should be honored when enabled: [test grid](https://testgrid.k8s.io/sig-apps#gce)
+- StatefulSet AvailableReplicas should get updated accordingly when MinReadySeconds is enabled: [test grid](https://testgrid.k8s.io/sig-apps#gce)
+
```

Comment:
Comments for lines that I can't comment on:

L356: "we will set `minReadySeconds` to 0" - who exactly is "we"? Do we really set it?
  Or do we implicitly default it on the controller side?

L357: "The default value of AvailableReplicas would be value of existing field in Status sub-resource called ReadyReplicas" - do we really default that field? This seems wrong.
It's just controller should start setting this field (together with other status fields).

L362: kube-apiserver will not clear automatically the field; it may be cleared if someone overwrites the object (not using patch). What happens in practice?


L431: please link?
[also - are they testing enablement/disablement (so switching on/off the feature in the middle of the test) or rather how it works with feature enabled and separately disabled]

Reply:
> L356: "we will set minReadySeconds to 0" - who exactly is "we"? Do we really set it?

minReadySeconds is an int32 whose default value is 0. We don't explicitly set it in [defaults.go](https://github.com/kubernetes/kubernetes/pull/100842/commits/49af83585220d815a1a29c01fde383d663c74fb7#diff-58ecd926004e524774a14cbf4e00cd64390bb8380a42c82b426a6f216e7dd612R30), if that is what you're asking 

> "The default value of AvailableReplicas would be value of existing field in Status sub-resource called ReadyReplicas" - do we really default that field? This seems wrong.

I think by default you mean the value which will be present in defaults.go. No, we don't set the value there.
If minReadySeconds is 0, the [default behavior of the controller ](https://github.com/kubernetes/kubernetes/pull/101316/files#diff-a1e25eb50e4b2aed8947c2caa5d6d1ddfef1a5f2f95fdf71f2388c657823ee13R316)is to set all the ready replicas as available replicas.

> L362: kube-apiserver will not clear automatically the field; it may be cleared if someone overwrites the object (not using patch). What happens in practice?

The values continue to [exist](https://github.com/kubernetes/kubernetes/pull/100842/commits/f3ab9550180a57e693105f77e8eba1fa00499fe2#r638225839) but won't be honored by the controller. I updated the statement to reflect it.


>  [also - are they testing enablement/disablement (so switching on/off the feature in the middle of the test) or rather how it works with feature enabled and separately disabled]

Not in the middle of tests but run test with feature enabled and then run the same test when the feature is disabled. Added the same in test plan section. Do you want anything else to be added?

Reply:
> I think by default you mean the value which will be present in defaults.go. No, we don't set the value there.
If minReadySeconds is 0, the default behavior of the controller is to set all the ready replicas as available replicas.

Please adjust then to say that "it's the controller that sets this field to the same value as ReadyReplicas.
This is different then defaulting (e.g. it's asynchronous).


> Not in the middle of tests but run test with feature enabled and then run the same test when the feature is disabled. Added the same in test plan section. Do you want anything else to be added?

Which is not really the enablement/disablement - we're testing the feature on and off.
It's a bit too late to ask for it, so I won't ask for adding such tests now, but please clarify in the doc that you're not testing enablement/disablement, but rather the feature being on and off.

Reply:
> Please adjust then to say that "it's the controller that sets this field to the same value as ReadyReplicas.

I did that in the last commit.

>  It's a bit too late to ask for it, so I won't ask for adding such tests now, but please clarify in the doc that you're not testing enablement/disablement, but rather the feature being on and off.

Sure. I updated the wording. If you have an example of how to enable/disable, I can try fitting it into this release. I was following the doc at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version